compaction.js

```diff
@@ -163,16 +182,16 @@ function Compaction(log, onDone) {
 
   function savePersistentState(cb) {
     if (!unshiftedBlockBuf) {
-      loadUnshiftedBlock(function onUnshiftedBlockLoaded() {
-        saveIt.call(this)
-      })
+      loadUnshiftedBlock(saveIt)
     } else {
-      saveIt.call(this)
+      saveIt()
     }
 
     function saveIt() {
       persistentState.save(
         {
+          version,
+          truncateBlockIndex,
```

Comment:
one things I don't understand from reading the code only is that `truncateBlockIndex` is only set to `compactedBlockIndex` when `stop()` is called. It seems like it should also be set before running `saveIt()` or am I missing something?

Reply:
So `saveIt` is part of `savePersistentState` which is **different** to `persistentState.save()`. Maybe a rename could help clarify that. `savePersistentState` is more like "load unshifted block then save persistent state" but that's a very long name.

We only set `truncateBlockIndex` when all blocks have been compacted, because when we compact restarts from a crash (line 148) it will "truncateAndBeDone" if the `truncateBlockIndex` is set. If it's not set, it means we still have to compact something.

Reply:
Said another way, the state file is almost like a write-ahead-log, but it's a state file, not a log, so we need to keep `truncateBlockIndex` at `0xFFFFFFFF` the whole time *until* truncation is about to be done, then we save the real truncate block index just before performing truncate. Then, we delete the whole file because we're fully done.

Reply:
Ahh, my confusion was actually the stop name. I was thinking this was something you could call at any point to stop compaction, but I now see that it means compaction is done :)